Fix wyverns disappearing on certain types of zoning

DiscipleOfEris opened the pull request:
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)
- [x] I've _**tested my code and the things my code has changed**_ since the last commit in the PR, and will test after any later commits

## What does this pull request do?

Wyverns, Automatons, and Avatars weren't being respawned when leaving town (if you entered with it) or when zoning via the `luautils::setPos()` function, which includes NPC interactions and the `!zone` command. They should now be retained in those instances.

Also, Jug Pets were being kept through normal zoning, but I've been informed that is not retail behavior so it will be removed by this PR as well.

Fixes #2536 
Fixes #2535 

## Steps to test these changes

See #2536 

More information on how this is handled by retail for different types of pets and different types of zoning would be valuable.

DiscipleOfEris committed:
Fix wyverns disappearing on certain types of zoning

Wyverns (and other pets retained through zoning) will not be respawned when leaving town (if you entered with it) and when zoning via the `luautils::setPos()` function, which includes NPC interactions and the `!zone` command.

claywar commented on src/map/lua/lua_baseentity.cpp:
```diff
@@ -2668,6 +2668,11 @@ void CLuaBaseEntity::setPos(sol::variadic_args va)
                 return;
             }
 
+            if (((CCharEntity*)m_PBaseEntity)->PPet != nullptr)
```
Just for consistency, can the C-style casts please be converted to static_cast?

`if (static_cast<CCharEntity*>(m_PBaseEntity)->PPet != nullptr)`

DiscipleOfEris replied:
Done. Touched up the code in the pet respawning section while at it, just so the block all matches.

DiscipleOfEris committed:
Switching to static_cast

Updating to the preferred static casting since we're touching this code anyway.

claywar commented on src/map/lua/lua_baseentity.cpp:
```diff
@@ -2668,17 +2668,22 @@ void CLuaBaseEntity::setPos(sol::variadic_args va)
-            ((CCharEntity*)m_PBaseEntity)->clearPacketList();
-            charutils::SendToZone((CCharEntity*)m_PBaseEntity, 2, zoneutils::GetZoneIPP(m_PBaseEntity->loc.destination));
-            //((CCharEntity*)m_PBaseEntity)->loc.zone->DecreaseZoneCounter(((CCharEntity*)m_PBaseEntity));
+            if (static_cast<CCharEntity*>(m_PBaseEntity)->PPet != nullptr)
```
Looking at this now, we should probably follow the example used in the upper block, and not cast each time, but instead declare PChar in this one as well (`auto* PChar = static_cast<CCharEntity*>(m_PBaseEntity);`)  That'll reduce the bulk of those repeated casts, and just replace with the new var.

WinterSolstice8 commented on src/map/lua/lua_baseentity.cpp:
```diff
@@ -2668,17 +2668,22 @@ void CLuaBaseEntity::setPos(sol::variadic_args va)
-            ((CCharEntity*)m_PBaseEntity)->clearPacketList();
-            charutils::SendToZone((CCharEntity*)m_PBaseEntity, 2, zoneutils::GetZoneIPP(m_PBaseEntity->loc.destination));
-            //((CCharEntity*)m_PBaseEntity)->loc.zone->DecreaseZoneCounter(((CCharEntity*)m_PBaseEntity));
+            if (static_cast<CCharEntity*>(m_PBaseEntity)->PPet != nullptr)
```
Should be `dynamic_cast<CPetEntity*>(m_PBaseEntity)->PPet != nullptr`
PPet will never be a character but could be a mob or a pet, and we only care if pets can zone. To the best of my knowledge this applies to all "real" pets, e.g. avatar/wyvern/automaton but unsure about call beast. I can test this real quick and will report back on if we need to check pet type for jug pets

dynamic_cast is done at runtime, which is important if there's a distinction between pets and mobs like this instance

WinterSolstice8 replied:
Wait, my brain isn't working from just getting home. I think Clay's comment is right.

DiscipleOfEris committed:
DRY compliance

Cache the casted variable instead of repetitively casting.

WinterSolstice8 commented:
Jug pets don't persist between zone, reverifying Avatars

DiscipleOfEris committed:
CPP Formatting compliance

WinterSolstice8 commented:
Avatars persist on zone too

DiscipleOfEris committed:
Jug pets should not persist

As per @WinterSolstice8, jug pets do not persist between zones on retail.

WinterSolstice8 reviewed (approved):
Looks good, if this still works after your edits this should be fine to merge

DiscipleOfEris commented:
> Avatars persist on zone too

Is there a difference in pet persistence between the different methods of zoning? Zoning normally, zoning via an NPC event, zoning into town and then out, etc.

DiscipleOfEris commented:
@WinterSolstice8 See #2536 where I suggest the types of things to check retail.

WinterSolstice8 commented:
test was: zone to south gustaberg, summon carby, go back in bastok, zone back out, carby still there. Job change (or subjob) should remove the pet, but not simply zoning into MH. HP warping elsewhere (to sky) carby still showed up. Same behavior as DRG wyvern/automaton. Essentially once they're out they persist until job/subjob change or death

DiscipleOfEris committed:
Avatar should persist between zoning

As per testing by @WinterSolstice8, avatars should also be maintained through the different types of zoning.

DiscipleOfEris commented:
Doesn't seem to be working now. Let me test.

TeoTwawki commented:
> Jug pets don't persist between zone, reverifying Avatars

_cries in beastmaster's gil_

DiscipleOfEris commented:
How many ticks of perpetuation cost are paid while zoning with an avatar? And is this the same as the number of regen/refresh ticks?

WinterSolstice8 commented:
None is accumulated when zoning back in, cost appears to only be deducted on tick when it's out in the zone

TeoTwawki commented:
> How many ticks of perpetuation cost are paid while zoning with an avatar? And is this the same as the number of regen/refresh ticks?

That will depend on how long the zoneout processing takes. Your character is still "active" on retail for a few sec on the black screen before the zone in starts. At the zone you are arriving in nothing should happen/count till the fade in, where char is already loaded and stats exist. To do this right we won't want to try any kind of timer - this is going to be variable lag depending, so we really just want to make sure we're not checking till after the char is loaded again

DiscipleOfEris committed:
Avatar should persist through zoning

We already fixed the Avatar check for respawning the avatar, now we need to check for storing it. Also clean up JUG_PET at the same time, as they're not supposed to be saved between zones.

DiscipleOfEris commented:
Okay. This PR also has the (intended) side effect of stopping Jug Pets from zoning with you. Avatars should be maintained through zoning now.
